ui/analyse/src/view/components.ts

```diff
@@ -1,84 +1,178 @@
-import { render as trainingView } from './roundTraining';
 import { view as actionMenu } from './actionMenu';
-import renderClocks from './clocks';
-import * as control from '../control';
-import crazyView from '../crazy/crazyView';
-import AnalyseCtrl from '../ctrl';
+import retroView from '../retrospect/retroView';
+import practiceView from '../practice/practiceView';
 import explorerView from '../explorer/explorerView';
-import forecastView from '../forecast/forecastView';
 import { view as forkView } from '../fork';
-import * as gridHacks from './gridHacks';
+import renderClocks from './clocks';
+import * as control from '../control';
 import * as chessground from '../ground';
+import AnalyseCtrl from '../ctrl';
 import { ConcealOf } from '../interfaces';
-import { view as keyboardView } from '../keyboard';
 import * as pgnExport from '../pgnExport';
-import retroView from '../retrospect/retroView';
-import practiceView from '../practice/practiceView';
-import serverSideUnderboard from '../serverSideUnderboard';
-import { render as renderTreeView } from '../treeView/treeView';
 import { spinnerVdom as spinner } from 'common/spinner';
+import * as Prefs from 'common/prefs';
+import statusView from 'game/view/status';
 import { stepwiseScroll } from 'common/scroll';
-import type * as studyDeps from '../study/studyDeps';
 import { renderNextChapter } from '../study/nextChapter';
-import * as Prefs from 'common/prefs';
-import StudyCtrl from '../study/studyCtrl';
+import { render as renderTreeView } from '../treeView/treeView';
+import * as gridHacks from './gridHacks';
 import { dispatchChessgroundResize } from 'common/resize';
-import { tourSide } from '../study/relay/relayTourView';
+import serverSideUnderboard from '../serverSideUnderboard';
 
-window.addEventListener('popstate', () => window.location.reload());
+import StudyCtrl from '../study/studyCtrl';
+import RelayCtrl from '../study/relay/relayCtrl';
+import type * as studyDeps from '../study/studyDeps';
 
-function makeConcealOf(ctrl: AnalyseCtrl): ConcealOf | undefined {
-  const conceal =
-    ctrl.study && ctrl.study.data.chapter.conceal !== undefined
-      ? {
-          owner: ctrl.study.isChapterOwner(),
-          ply: ctrl.study.data.chapter.conceal,
-        }
-      : null;
-  if (conceal)
-    return (isMainline: boolean) => (path: Tree.Path, node: Tree.Node) => {
-      if (!conceal || (isMainline && conceal.ply >= node.ply)) return null;
-      if (treePath.contains(ctrl.path, path)) return null;
-      return conceal.owner ? 'conceal' : 'hide';
-    };
-  return undefined;
+export interface ViewContext {
+  ctrl: AnalyseCtrl;
+  deps?: typeof studyDeps;
+  study?: StudyCtrl;
+  relay?: RelayCtrl;
+  concealOf?: ConcealOf;
+  showCevalPvs: boolean;
+  menuIsOpen: boolean;
+  gamebookPlayView?: VNode;
+  playerBars: VNode[] | undefined;
+  playerStrips: [VNode, VNode] | undefined;
+  gaugeOn: boolean;
+  needsInnerCoords: boolean;
+  tourUi?: VNode;
 }
```

Comment:
That's useful but I'm concerned that it leads to using more typescript `!`, like `ctx.study!`. This would negate the benefits we get from statically checking that we do have a `study` (or `relay` or anything).

So this ViewContext is good in my opinion, at the condition that we don't do things like this: 
```ts
function(ctx: ViewContext) { ctx.study!.something() }
```

but instead keep on passing the study that the compiler knows exists like so:
```ts
function(ctx: ViewContext, study: StudyCtrl) { study.something() }
```

Does that make sense?